File: python/src/problems101_200/valid_palindrome.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def is_palindrome(self, s: str) -> bool:
        pass


class ReverseString(Problem):
    @override
    def is_palindrome(self, s: str) -> bool:
        builder: str = ''
        for c in s:
            if c.isalnum():
                builder += c

        return builder.lower() == builder.lower()[::-1]


class CompareHalfString(Problem):
    @override
    def is_palindrome(self, s: str) -> bool:
        builder: str = ''
        for c in s:
            if c.isalnum():
                builder += c

        s: str = builder
        start: int = 0
        end: int = len(s) - 1
        while start < end:
            if s[start].lower() != s[end].lower():
                return False
            start += 1
            end -= 1
        return True


SOLUTIONS: list[Problem] = [ReverseString(), CompareHalfString()]

File: typescript/src/problems101_200/valid-palindrome.ts
class Problem {
  isPalindrome(s: string): boolean {
    throw new Error(`Not implemented (${s}).`);
  }
}

class BruteForce extends Problem {
  isPalindrome(s: string): boolean {
    let builder = '';
    for (const c of s) {
      if (/[a-zA-Z0-9]/.test(c)) {
        builder += c;
      }
    }
    s = builder.toLowerCase();
    return s === s.split('').reverse().join('');
  }
}

class CompareHalfString extends Problem {
  isPalindrome(s: string): boolean {
    let builder = '';
    for (const c of s) {
      if (/[a-zA-Z0-9]/.test(c)) {
        builder += c;
      }
    }
    s = builder;
    let start = 0;
    let end = s.length - 1;
    while (start < end) {
      if (s[start++].toLowerCase() !==
          s[end--].toLowerCase()) {
        return false;
      }
    }
    return true;
  }
}

const SOLUTIONS = [new BruteForce(), new CompareHalfString()];
export default SOLUTIONS;
